From: Steve Langasek <steve.langasek@linaro.org>
Date: Fri, 5 Jun 2026 17:15:23 +0800
Subject: make ldconfig use the built-in system paths

Bug-Ubuntu: https://bugs.launchpad.net/bugs/736932

Make ldconfig use the same view of built-in system paths that ld.so does,
instead of just using SLIBDIR and LIBDIR; this corrects a failure of
ldconfig to cache libraries in non-multiarch directories when building for
multiarch, even though they're on the system path.
---
 elf/ldconfig.c | 24 +++++++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/elf/ldconfig.c b/elf/ldconfig.c
index aca9674..ea48de6 100644
--- a/elf/ldconfig.c
+++ b/elf/ldconfig.c
@@ -55,6 +55,17 @@
 
 #define PACKAGE _libc_intl_domainname
 
+/* Get the generated information about the trusted/standard directories.  */
+#include "trusted-dirs.h"
+
+static const char system_dirs[] = SYSTEM_DIRS;
+static const size_t system_dirs_len[] =
+{
+  SYSTEM_DIRS_LEN
+};
+#define nsystem_dirs_len \
+  (sizeof (system_dirs_len) / sizeof (system_dirs_len[0]))
+
 /* List of directories to handle.  */
 struct dir_entry
 {
@@ -1285,12 +1296,19 @@ main (int argc, char **argv)
 
   if (!opt_only_cline)
     {
+      const char *strp = system_dirs;
+      size_t idx = 0;
+
       parse_conf (config_file, true);
 
       /* Always add the standard search paths.  */
-      add_system_dir (SLIBDIR);
-      if (strcmp (SLIBDIR, LIBDIR))
-	add_system_dir (LIBDIR);
+      do
+        {
+          add_system_dir (strp);
+          strp += system_dirs_len[idx] + 1;
+          idx++;
+        }
+      while (idx < nsystem_dirs_len);
     }
 
   const char *aux_cache_file = _PATH_LDCONFIG_AUX_CACHE;
